### PR TITLE
Metadata 2.2 and 2.3

### LINKED ISSUE #9660: Support Metadata Version 2.2
**What's the problem this feature will solve?**
Since PEP 643 is accepted, PyPI should start accepting Metadata 2.2 distributions. See  PyO3/maturin#564.

**Describe the solution you'd like**
Allow `Metadata-Version: 2.2` to pass the legacy upload endpoint’s form validation, and validate the corresponding `Dynamic` field.

The latter part could be a bit tricky since `Dynamic` is not allowed for wheels. Not sure how it could be best handled (I don’t have much experience with WTForm).

**Additional context**
https://www.python.org/dev/peps/pep-0643/

### LINKED ISSUE #11526: Add support for PEP 685 (Metadata Version 2.3)
**What's the problem this feature will solve?**
[PEP 685](https://peps.python.org/pep-0685/) specifies rules for normalisation of extra names in core metadata, and in turn bumps the metadata version to 2.3. Currently it is not possible to upload wheels with metadata 2.3 because warehouse rejects them as having an unsupported metadata version.

**Describe the solution you'd like**
The PEP suggests tools consuming metadata (like warehouse) emit warnings if invalid extra names are found for *older* metadata versions, although the tool may raise an error. I assume there's no mechanism in the upload API to return a warning, only an error? I don't think an error is appropriate here as it will likely break uploads of packages with previously valid metadata.

Tools should also warn (but may error) for invalid extras with metadata version 2.3. 

Therefore I propose leaving the validation unchanged for older versions, but returning an error if the metadata version is 2.3 (or later) and an extra name is invalid.

**Additional context**

I have an initial implementation here: https://github.com/pypa/warehouse/pull/11380/commits/48b5063a65a382d3bba622351858aea9c058f6c0

### PR DESCRIPTION
**Pull request by konstin**
This is an attempt to finish #11380 and unblock uploading metadata version 2.2 and 2.3, which i'd like to have for maturin and my resolver prototype. I've applied all review comments and added a check that dynamic isn't used <2.2.

Fixes #9660, fixes #11526, closes #11380.

### CODE REVIEW

File: warehouse/packaging/models.py
Comment (ewdurbin) [inline warehouse/packaging/models.py:418]: I don't think we need this to be a relation. I'm going to push a change to convert it to an Array containing the Enum

**Review by ewdurbin — DISMISSED** [review]
The only thing I would want to see to call this shippable is a test that exercises the storage/retrieval of Dynamic fields.

I think this could be added to `tests.unit.forklift.test_legacy:test_successful_upload` or maybe more favorably built out as an additional test `tests.unit.forklift.test_legacy:test_successful_upload_metadata_2_3`.

Just some assertions of the resulting value of `Release.dynamic`

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [inline warehouse/forklift/legacy.py:407]: From https://github.com/pypi/warehouse/pull/11380#discussion_r1033894111:

PEP 643 only says:

> The fields `Name` and `Version` MUST NOT be marked as Dynamic.

Should `Metadata-Version` be included here? Or should the PEP be updated?

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [inline warehouse/forklift/legacy.py:607]: From https://github.com/pypi/warehouse/pull/11380#discussion_r1033923344

This needs to be added as a field of the Release model as well: https://github.com/pypi/warehouse/blob/main/warehouse/packaging/models.py#L423-L604

**Review by ewdurbin — COMMENTED** [review]
Ported outstanding comments from #11380

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [reply warehouse/forklift/legacy.py:607]: I think this can be a generic `sqlalchemy.dialects.postgresql.ARRAY` of `sqlalchemy.Text` since the incoming are validated:

```
    provides_extra = Column(ARRAY(Text))
```

**Comment by ewdurbin** [comment]
> The only thing I would want to see to call this shippable is a test that exercises the storage/retrieval of Dynamic fields.
> 
> I think this could be added to `tests.unit.forklift.test_legacy:test_successful_upload` or maybe more favorably built out as an additional test `tests.unit.forklift.test_legacy:test_successful_upload_metadata_2_3`.
> 
> Just some assertions of the resulting value of `Release.dynamic`

Missed that Provides-Extra also needs to be stored. Similarly a test for storage/retrieval of that data would be great to see.

File: warehouse/forklift/legacy.py
Comment (pfmoore) [reply warehouse/forklift/legacy.py:407]: This is already the case in [the spec](https://packaging.python.org/en/latest/specifications/core-metadata/#dynamic-multiple-use). It was added in https://github.com/pypa/packaging.python.org/pull/1057 (I can't check this right now as github is playing up, but that's what git blame tells me locally!)

Comment (pfmoore) [reply warehouse/forklift/legacy.py:407]: The principle is that the spec is the definitive document, and the PEP is historical, so there's no need to update the PEP.

Comment (ewdurbin) [reply warehouse/forklift/legacy.py:407]: I'll defer to that!

**Comment by dstufft** [comment]
We probably want to expose this value in the existing JSON API, XMLRPC API, and.. .maybe in the UI? Not sure.

**Comment by ewdurbin** [comment]
Hi @konstin, wanted to check in and see if you planned to continue work on this. If not I'm happy to pick it up and continue on with this PR.

**Comment by ewdurbin** [comment]
> We probably want to expose this value in the existing JSON API, XMLRPC API, and.. .maybe in the UI? Not sure.

Which value, Provides-Extra?

**Comment by dstufft** [comment]
`Dynamic`, and whatever other fields are new

**Comment by ewdurbin** [comment]
> We probably want to expose this value in the existing JSON API, XMLRPC API, and.. .maybe in the UI? Not sure.

I will implement this in JSON API, but won't touch XMLRPC. There's nowhere we would expose this on the XMLRPC that isn't currently slated for deprecation. If people want programatic access to this data, they should retrieve it from JSON.

I'll give a _smol_ attempt to render it in the web UI, but consider it beyond scope of this PR.

**Comment by ewdurbin** [comment]
Low budget display if available on project pages:

low budget

**Comment by ewdurbin** [comment]
@di @dstufft @miketheman, this is good for a review pass.

**Comment by konstin** [comment]
Sorry for responding so slow and thank you for addressing all the review comments!

**Comment by konstin** [comment]
Should i rebase this?

**Comment by di** [comment]
@konstin Yes please!

**Comment by konstin** [comment]
rebased and updated with the changes from main

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:418]:
```suggestion
    declared_dynamic_fields = set(map(str.title, field.data or []))
    declared_dynamic_fields = {str.title(k) for k in field.data}
```

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:428]: I think what we actually want to do here is check if all the values in `Dynamic` are present as other metadata fields -- other verification checks will ensure that these fields are valid metadata fields for that metadata version. I'm not sure we need the enum at all.

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:441]: This should already be getting validated elsewhere (and before this function is called) so it should be safe to assume that the metadata version is valid.

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:626]:
```suggestion
        validators=[wtforms.validators.Optional(), _validate_provides_extras],
```

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:434]:
```suggestion
def _validate_provides_extras(form, field):
```

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:678]:
```suggestion
        metadata_version = packaging.version.Version(form.metadata_version.data)
        if metadata_version >= packaging.version.Version("2.2")
```

File: warehouse/legacy/api/json.py
Comment (di) [inline warehouse/legacy/api/json.py:160]: Should we make a distinction here between when the field is not provided, and when it is provided but empty? I'm not sure if the PEP addresses this or makes a distinction.

File: warehouse/legacy/api/json.py
Comment (di) [inline warehouse/legacy/api/json.py:167]: Ditto here.

File: warehouse/legacy/api/json.py
Comment (konstin) [reply warehouse/legacy/api/json.py:160]: Both `Provides-Extra` and `Dynamic` are optional multiple use fields, is at least `METADATA` files can't express a difference between missing-empty and present-empty, so i'd say it's fine to treat them the same here.

File: warehouse/forklift/legacy.py
Comment (konstin) [reply warehouse/forklift/legacy.py:678]: Note that this means that we allow 3.0 here (i deleted the test for that from the old PR)

**Comment by konstin** [comment]
I tried to address all review comments, i had to change some of the suggested changes though to make it work

**Comment by miketheman** [comment]
> Is there a way to tell the linters and formatters (or the `make` commands in general) to not lint my local `.venv`?

Tough to say exactly, since I don't use the same .venv style, as most development happens inside a container. You might want to add a .venv directory entry to https://github.com/pypi/warehouse/blob/main/.dockerignore so that it's not added to the container - give that a shot and if it works, feel free to add that to its own PR.

**Comment by konstin** [comment]
Thanks, created https://github.com/pypi/warehouse/pull/14739

**Comment by brettcannon** [comment]
Anything that can be done to help progress this? Metadata version 2.2 and newer are important for resolvers/installers as it lets them (potentially) skip having to download and build an sdist -- if the `PKG-INFO` is made available separately -- just to get the list of dependencies which is a massive compute, time, and bandwidth saver.

**Comment by ewdurbin** [comment]
Rebased and ready for another review. I think the only outstanding questions are regarding how None versus empty are presented in JSON API: https://github.com/pypi/warehouse/pull/13606#discussion_r1261300580

@brettcannon if you could weigh in there I think we can probably move forward.

File: warehouse/legacy/api/json.py
Comment (pfmoore) [reply warehouse/legacy/api/json.py:160]: I agree with @konstin - there's no meaningful way of interpreting "present but empty" *except* by not including the metadata item in the `METADATA` file, so the two are for all practical purposes equivalent.

File: warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py
Comment (di) [inline warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py:125]: This migration should probably happen first, in a separate PR. It probably should also set a higher statement timeout than the default.

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:405]:
```suggestion
    # "Supported-Platform" is missing in the metadata form since it does not have any
    # constraints
```

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:408]: I understand why this is here but this seems a bit weird, perhaps we should add the field to the form instead, even if we don't do anything with it.

It also seems like a bit of a hack that we're relying on the description for these values, should we be getting this from `DynamicFieldsEnum` instead?

File: warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py
Comment (konstin) [reply warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py:125]: What lock/statement timeouts do you suggest?

Comment (di) [reply warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py:125]: Adding this at the top of the migration should be sufficient:

```python
    conn = op.get_bind()
    conn.execute(sa.text("SET statement_timeout = 120000"))
```

File: warehouse/legacy/api/json.py
Comment (brettcannon) [reply warehouse/legacy/api/json.py:160]: Since both results are functionally the same I think either approach is fine.

**Comment by brettcannon** [comment]
> @brettcannon if you could weigh in there I think we can probably move forward.

Done!

File: warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py
Comment (konstin) [reply warehouse/migrations/versions/778f1c01a019_release_add_provides_extra_and_dynamic_.py:129]: I've created https://github.com/pypi/warehouse/pull/15452, i unfortunately can't retarget this PR onto the other branch since the branch is in my fork

File: warehouse/forklift/legacy.py
Comment (konstin) [reply warehouse/forklift/legacy.py]: > I understand why this is here but this seems a bit weird, perhaps we should add the field to the form instead, even if we don't do anything with it.

Done

> It also seems like a bit of a hack that we're relying on the description for these values, should we be getting this from DynamicFieldsEnum instead?

I think we had that and changed it due to https://github.com/pypi/warehouse/pull/13606#discussion_r1261291228, or am i misreading your comment?

File: warehouse/forklift/legacy.py
Comment (di) [reply warehouse/forklift/legacy.py]: What I meant by this was whether this check should ensure that all the fields listed in `Dynamic` are actually provided as metadata -- e.g., if we have `Dynamic: Author` but `Author` is not provided, should this fail? I think the PEP is unclear on this.

File: warehouse/forklift/legacy.py
Comment (di) [reply warehouse/forklift/legacy.py]: What I meant there is that I don't think we don't need to check if `Dynamic` contains fields that aren't valid metadata fields (e.g. `Dynamic: ThisDoesntExist`), as those will fail when they're provided as metadata fields (e.g. `ThisDoesntExist: Foo`) *but* that assumes that the validator will always fail when `Dynamic` contains any value that isn't also present as a valid metadata field (e.g., `Dynamic: Author`, but `Author` isn't provided).

If the latter is true, we don't need the Enum or to check if `Dynamic` contains a valid metadata field name, we only need to:

1. Check that it doesn't contain a disallowed field name (`Dynamic: Metadata-Version` should fail);
2. Check that it only contains field names that are also provided as metadata (`Dynamic: Author` should fail if `Author` is not provided, `Dynamic: ThisDoesntExist` should fail because the form prevents `ThisDoesntExist` from being provided).

File: warehouse/forklift/legacy.py
Comment (konstin) [reply warehouse/forklift/legacy.py]: I think both cases allowed, e.g. if you have `Dynamic: Requires-Dist`, you may not provide any requirements, but you still have to set it as `Dynamic` or consumers can assume it's a package with no dependencies. But you may give an incomplete/insufficiently constrained list and tighten it at build time (i have been told this is important for some packages that can build for different numpy version).

File: warehouse/forklift/legacy.py
Comment (di) [reply warehouse/forklift/legacy.py]: OK, that makes sense too! Given that I think we do need the Enum but we can simplify this, I'll provide a suggestion in a moment.

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:410]:
```suggestion
    declared_dynamic_fields = {str.title(k) for k in field.data or []}
    disallowed_dynamic_fields = {"Name", "Version", "Metadata-Version"}
    if invalid := (declared_dynamic_fields & disallowed_dynamic_fields):
        raise wtforms.validators.ValidationError(
            f"The following metadata field(s) are valid, but cannot be marked as dynamic: {invalid!r}",
        )
    allowed_dynamic_fields = set(DynamicFieldsEnum.enums)
    if invalid := (declared_dynamic_fields - allowed_dynamic_fields):
        raise wtforms.validators.ValidationError(
            f"The following metadata field(s) are not valid and cannot be marked as dynamic: {invalid!r}"
        )
```
I think it's better to use an explicit Enum for this, because it's not actually guaranteed that `description` is set for all these fields. Also, ideally, we would reuse that same enum in the migration as well.

Comment (konstin) [reply warehouse/forklift/legacy.py]: If we import the enum in the migration, wouldn't this break the migration if we ever update the enum? (Sorry if that's a stupid question, i don't have any experience with alembic)

Comment (di) [reply warehouse/forklift/legacy.py]: It's a good question. It wouldn't necessarily break the migration itself (because it would always pull the latest enum when we were running the migration), but it might make the migration technically incorrect if that were ever to happen. I'm OK with not using the enum -- we can verify that they are the same before merging.

Comment (konstin) [reply warehouse/forklift/legacy.py]: With that resolved, is there anything else you need to be changed to be able to merge #15452?

Comment (di) [reply warehouse/forklift/legacy.py]: No, I think its fine!

**Review by di — APPROVED** [review]
LGTM, is blocked on merging #15452 first.

**Review by ewdurbin — APPROVED** [review]

**Pull request merged**
